### ISSUE TITLE
Diagnostics inline message not going away after fix

### ISSUE THREAD

**Issue by jader-rubini** [zed-industries/zed#16135]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I noticed in the last few days that my inline diagnostics messages aren't going away anymore after I apply the fixes, not even after I save the file again. The only way to make them go is closing the buffer and reopening again, which is not a great experience when you're working iteratively.

<img width="686" alt="Screenshot 2024-08-12 at 18 30 44" src="https://github.com/user-attachments/assets/3cf1d0f3-52ae-4044-8957-e6846a9f1d26">

### Environment

Zed: v0.147.2 (Zed)
OS: macOS 14.5.0
Memory: 16 GiB
Architecture: aarch64

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by dlants** [comment]
I'm experiencing the same thing with typescript lsp.

Zed 0.160.7
MacOS 14.6.1 (Apple M1 Pro)

<img width="792" alt="Screenshot 2024-11-11 at 10 01 08 PM" src="https://github.com/user-attachments/assets/76b9e0df-31ea-436b-a592-2cd206ecd7d4">

**Comment by bofm** [comment]
Here is my current workaround for this:

Config:
```json
{
    "context": "Editor",
    "bindings": {
       "cmd-s": ["workspace::SendKeystrokes", "cmd-s escape"]
    }
}
```

1. So, let's say there is an file with an error and a diagnostic message showing inline after `editor::GoToDiagnostic` command.
2. I fix the error.
3. I press cmd-s to save the file. All diagnostic inline messages go away (both fixed the unfixed ones).
4. I have to press a hotkey assigned to `editor::GoToDiagnostic` again to move on.

**Comment by bofm** [comment]
I also tried to do `"cmd-s": ["workspace::SendKeystrokes", "cmd-s escape alt-o"]` where `alt-o` is bound to `editor::GoToDiagnostic`. But it didn't work because the key combination hits much faster than the language server rechecks and clears the diagnostic error for the fixed code. Some kind of delay before re-running the `editor::GoToDiagnostic` is required.

**Comment by bofm** [comment]
The inline diagnostic messages are now disappear after save.

**Comment by mikalsande** [comment]
I agree with @bofm, this looks like its fixed now. The diagnostic disappears when navigating away from it and when saving.


**Environment**
Zed: 0.201.8
OS: macOS 15.6.1
Memory: 36 GiB
Architecture: aarch64

Have tested on with Rust and Go with these examples.

Rust
```rust
use std::convert::From;

#[derive(Debug)]
struct Number {
    value: i32,
}

impl From<i32> for Number {
    fn from(item: i32) -> Self {
        Number { value: item }
    }
}

fn main() {
    let num = Number::from("30");
    println!("My number is {:?}", num);
}
```

Go
```go
package main

import "fmt"

func intSeq() func() int32 {
	i := 0
	return func() int {
		i++
		return i
	}
}

func main() {
	nextInt := intSeq()

	fmt.Println(nextInt())
	fmt.Println(nextInt())
	fmt.Println(nextInt())

	newInts := intSeq()
	fmt.Println(newInts())
}
```

Tested by doing
- Open the file, go to the error with `editor: go to diagnostic`.
- Fix the error and save.
- The diagnostic disappears, there are different behaviors, my guess is that this is because the LSPs operate differently. Both disappear when fixing the error though
    - For Go it disappears as soon as I start removing 32 from int32.
    - For Rust it disappears after saving.

@jader-rubini Can you retest?
